cudax/include/cuda/experimental/__device/all_devices.cuh

```diff
@@ -0,0 +1,244 @@
+  // `cudaGetDeviceCount`.
+  mutable _CUDA_VSTD::atomic<size_type> __size_{~0UL};
+};
+
+//! @brief A random-access iterator over all available CUDA devices
+struct all_devices::iterator
+{
+  using value_type        = device;
+  using difference_type   = int;
+  using pointer           = const device*;
+  using reference         = device;
+  using iterator_category = ::cuda::std::random_access_iterator_tag;
+  using iterator_concept  = ::cuda::std::random_access_iterator_tag;
+
+  //! @brief Proxy object used as the return value of `operator->`
+  struct __proxy
+  {
+    _CCCL_NODISCARD const device* operator->() const noexcept
+    {
+      return &__dev_;
+    }
+
+    device __dev_;
+  };
+
+  _CCCL_NODISCARD constexpr device operator*() const noexcept
+  {
+    return device(__id_);
+  }
+
+  _CCCL_NODISCARD constexpr device operator[](int __offset) const noexcept
+  {
+    return device(__id_ + __offset);
+  }
+
+  _CCCL_NODISCARD constexpr __proxy operator->() const noexcept
+  {
+    return __proxy{device(__id_)};
+  }
+
+  constexpr iterator& operator++() noexcept
+  {
+    ++__id_;
+    return *this;
+  }
+
+  constexpr iterator operator++(int) noexcept
+  {
+    auto tmp = *this;
+    ++__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator--() noexcept
+  {
+    --__id_;
+    return *this;
+  }
+
+  constexpr iterator operator--(int) noexcept
+  {
+    auto tmp = *this;
+    --__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator+=(int __offset) noexcept
+  {
+    __id_ += __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator+(int __offset) noexcept
+  {
+    return iterator(__id_ + __offset);
+  }
+
+  constexpr iterator& operator-=(int __offset) noexcept
+  {
+    __id_ -= __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator-(int __offset) noexcept
+  {
+    return iterator(__id_ - __offset);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr iterator operator+(int __lhs, const iterator& __rhs) noexcept
+  {
+    return iterator(__lhs + __rhs.__id_);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr int operator-(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ - __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator==(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ == __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator!=(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ != __rhs.__id_;
+  }
+
+private:
+  friend struct all_devices;
+
+  constexpr explicit iterator(int __id) noexcept
+      : __id_(__id)
+  {}
+
+  int __id_ = 0;
+};
+
+_CCCL_NODISCARD inline constexpr device all_devices::operator[](size_type __id_) const noexcept
+{
+  assert(__id_ < size());
+  return device(static_cast<int>(__id_));
+}
+
+_CCCL_NODISCARD inline all_devices::size_type all_devices::size() const
```

Comment:
I find it scary that the user has to know that they must call `size()` before doing anything else

Reply:
they don't. everything else, like calls to `end()`, is expressed in terms of calls to the `size()` member.

Reply:
![download](https://github.com/user-attachments/assets/a4a4c218-9f62-47ef-837f-afe2299bf9e3)